### ISSUE TITLE
Is this block of code autoregressive?

### ISSUE THREAD

**Issue by Dovermore** [dtsip/in-context-learning#2]
Hello, I've been examining your code and academic work. I see that `_backbone` is a GPT-2 model from the `transformers` library. It behaves as autoregressive when the `past` parameter is active. 

Given assumptions like `batch_size=1`, `points=3`, and `n_dim=2`, `xs` will take the shape `(1,3,2)`, and `ys` (or `f(xs)` as in the paper) will have the shape `(1,3)`, aligned with the target outputs of the task. 

Here's the relevant block of code:
```python
@staticmethod
def _combine(xs_b, ys_b):
    """Interleaves the x's and the y's into a single sequence."""
    bsize, points, dim = xs_b.shape
    ys_b_wide = torch.cat(
        (
            ys_b.view(bsize, points, 1),
            torch.zeros(bsize, points, dim - 1, device=ys_b.device),
        ),
        axis=2,
    )
    zs = torch.stack((xs_b, ys_b_wide), dim=2)
    zs = zs.view(bsize, 2 * points, dim)
    return zs

def forward(self, xs, ys, inds=None):
    if inds is None:
        inds = torch.arange(ys.shape[1])
    else:
        inds = torch.tensor(inds)
        if max(inds) >= ys.shape[1] or min(inds) < 0:
            raise ValueError("inds contain indices where xs and ys are not defined")
    zs = self._combine(xs, ys)
    embeds = self._read_in(zs)
    output = self._backbone(inputs_embeds=embeds).last_hidden_state
    prediction = self._read_out(output)
    return prediction[:, ::2, 0][:, inds]
```

For `zs`, it seems to have a format of `[[x1, f(x1), x2, f(x2), x3, f(x3)]]`, implying `_backbone` has visibility into all values of `ys`, including the target `y3`. The model seems to be predicting `y1` through `y3` based on these inputs.

Though the last line in the `forward`, `prediction[:, ::2, 0][:, inds]` is selecting hidden states corresponding to all `xs` instead of `f(xs)`, the `f(xs)` are implicitly leaked into all the hidden states.

And in the `eval_batch` function, I don't observe any safeguards against data leakage. Here's the code for reference:
```python
def eval_batch(model, task_sampler, xs, xs_p=None):
    task = task_sampler()
    if torch.cuda.is_available() and model.name.split("_")[0] in ["gpt2", "lstm"]:
        device = "cuda"
    else:
        device = "cpu"

    if xs_p is None:
        ys = task.evaluate(xs)
        pred = model(xs.to(device), ys.to(device)).detach()
        metrics = task.get_metric()(pred.cpu(), ys)
    else:
        b_size, n_points, _ = xs.shape
        metrics = torch.zeros(b_size, n_points)
        for i in range(n_points):
            xs_comb = torch.cat((xs[:, :i, :], xs_p[:, i:, :]), dim=1)
            ys = task.evaluate(xs_comb)

            pred = model(xs_comb.to(device), ys.to(device), inds=[i]).detach()
            metrics[:, i] = task.get_metric()(pred.cpu(), ys)[:, i]
    return metrics
```
Specifically, the line `xs_comb = torch.cat((xs[:, :i, :], xs_p[:, i:, :]), dim=1)` and `ys = task.evaluate(xs_comb)` seem to suggest that `ys`, the query into the model, is composed of `(x1, f(x1), . . . , xk, f(xk), x_query, f(x_query)` instead of excluding `f(x_query)`. 

Could you clarify what I might be missing? Any insights would be greatly appreciated.

---

Best,
Calvin

**Comment by dtsip** [comment]
Hi, since the `_backbone` is auto-regressive, and each x_i/y_i is a separate token, the prediction for `y1` depends only on `x1`, the prediction for `y2` depends only on `x1, y1, x2`, and the prediction for `y3` depends only on `x1, y1, x2, y2, x3` which is exactly what we want.

**Comment by Dovermore** [comment]
Thanks, I think I identified the lines from GPT2 relating to causal inference,
```
if not self.is_cross_attention:
    # if only "normal" attention layer implements causal mask
    query_length, key_length = query.size(-2), key.size(-2)
    causal_mask = self.bias[:, :, key_length - query_length : key_length, :key_length].bool()
    attn_weights = torch.where(causal_mask, attn_weights, self.masked_bias.to(attn_weights.dtype))
```
